waspc/cli/src/Wasp/Cli/Command/News/Listing.hs

```diff
@@ -0,0 +1,117 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+
+module Wasp.Cli.Command.News.Listing
+  ( -- * News Listing
+
+    -- This modules captures all decisions Wasp has to make when showing Wasp
+    -- news under different contexts (i.e., who requested the news listing):
+    --   - Which news to show under which context.
+    --   - Which news to mark as seen.
+    --   - Whether the user should interactively confirm they have seen the
+    --     news.
+    --   - etc.
+    NewsListing (..),
+    getNewsToShow,
+    isConfirmationRequired,
+    getNewsToMarkAsSeen,
+    listNews,
+    shouldWaspListMustSeeNews,
+  )
+where
+
+import Control.Monad (unless, when)
+import Data.List (intercalate)
+import qualified Data.Time as T
+import Wasp.Cli.Command.News.Core (NewsEntry (..), NewsLevel (..))
+import Wasp.Cli.Command.News.Display (showNewsEntry)
+import Wasp.Cli.Command.News.LocalNewsState
+  ( LocalNewsState,
+    emptyLocalNewsState,
+    markNewsAsSeen,
+    saveLocalNewsState,
+    setLastListingTimestamp,
+    wasLastListingMoreThanNHoursAgo,
+    wasNewsEntrySeen,
+  )
+import Wasp.Cli.Interactive (askForConfirmationWithTimeout)
+import Wasp.Util.Terminal (styleCode)
+
+shouldWaspListMustSeeNews :: LocalNewsState -> IO Bool
+shouldWaspListMustSeeNews = wasLastListingMoreThanNHoursAgo 24
+
+data NewsListing
+  = UserListingAllNews {allNews :: [NewsEntry]}
+  | WaspListingMustSeeNews {allNews :: [NewsEntry]}
+  deriving (Show, Eq)
+
+getNewsToShow :: LocalNewsState -> NewsListing -> [NewsEntry]
+getNewsToShow localState = \case
+  UserListingAllNews {allNews} -> allNews
+  WaspListingMustSeeNews {allNews}
+    | doesUserHaveNewsHistory localState -> unseenNewsThatUserMustSee allNews
+    -- If the user has no news history, we don't want to hit them with all past news.
+    | otherwise -> []
+  where
+    unseenNewsThatUserMustSee = filter ((>= Important) . level) . filter isUnseen
+    isUnseen = not . wasNewsEntrySeen localState
+
+isConfirmationRequired :: LocalNewsState -> NewsListing -> Bool
+isConfirmationRequired localState listing = case listing of
+  UserListingAllNews {} -> False
+  WaspListingMustSeeNews {} -> any ((== Critical) . level) $ getNewsToShow localState listing
+
+getNewsToMarkAsSeen :: LocalNewsState -> NewsListing -> [NewsEntry]
+getNewsToMarkAsSeen localState listing = case listing of
+  UserListingAllNews {} -> getNewsToShow localState listing
```

Comment:
Hm. So I remember I had old comment here saying that we should not just return all the news, but only those that haven't been seen before, so the new ones.
And you did change this part, but to reurn all news to show, which are again all news, so the same behaviour but semantically different, and that might be ok if those sematnics made sense, but I don't see how they do, why would we mark as seen all the news that you will show? I guess because they are shown. Ok that is certianly an improvement over just marking all news, yeah I get it. But still hm. It still might again mark already marked news. If we are ok with that, then ok I guess.